File: Java/data_structures/disjoint_sets/DisjointSetUnion.java
import java.util.Arrays;

/**
 * Data Structure: Disjoint Set Union (Union-Find)
 * Description: Maintains a collection of disjoint sets with near O(1) amortized
 *              operations using path compression (find) and union by rank.
 *
 * Operations:
 *  - find(x): returns representative (root) of the set containing x
 *  - union(x, y): merges the sets containing x and y
 *  - connected(x, y): true if x and y are in the same set
 *
 * Time Complexity: Amortized ~O(1) per operation (inverse Ackermann)
 * Space Complexity: O(n)
 */
public class DisjointSetUnion {
    private final int[] parent;
    private final int[] rank;

    /**
     * Constructs a DSU for elements 0..n-1, each in its own set.
     * @param n number of elements
     * @throws IllegalArgumentException if n < 0
     */
    public DisjointSetUnion(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        this.parent = new int[n];
        this.rank = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    /**
     * Finds the representative of x with path compression.
     */
    public int find(int x) {
        if (x < 0 || x >= parent.length) throw new IndexOutOfBoundsException("index out of bounds: " + x);
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    /**
     * Unions the sets containing x and y using union by rank.
     * @return true if a merge happened, false if already in same set
     */
    public boolean union(int x, int y) {
        int rx = find(x);
        int ry = find(y);
        if (rx == ry) return false;
        if (rank[rx] < rank[ry]) {
            parent[rx] = ry;
        } else if (rank[rx] > rank[ry]) {
            parent[ry] = rx;
        } else {
            parent[ry] = rx;
            rank[rx]++;
        }
        return true;
    }

    /**
     * Checks if x and y belong to the same set.
     */
    public boolean connected(int x, int y) {
        return find(x) == find(y);
    }

    /**
     * Returns the number of elements managed by this DSU.
     */
    public int size() {
        return parent.length;
    }

    @Override
    public String toString() {
        // Show parent array for debugging
        return "parent=" + Arrays.toString(parent) + ", rank=" + Arrays.toString(rank);
    }

    public static void main(String[] args) {
        // Example usage
        DisjointSetUnion dsu = new DisjointSetUnion(10);
        System.out.println("Initial: " + dsu);

        dsu.union(0, 1);
        dsu.union(2, 3);
        dsu.union(3, 4);
        dsu.union(1, 2); // merges {0,1} with {2,3,4}
        dsu.union(5, 6);

        System.out.println("After unions: " + dsu);
        System.out.println("connected(0,4) -> " + dsu.connected(0, 4)); // true
        System.out.println("connected(0,6) -> " + dsu.connected(0, 6)); // false

        // Path compression check
        System.out.println("find(4) root -> " + dsu.find(4));
        System.out.println("Final: " + dsu);

        System.out.println("Ops: find/union amortized ~O(1); Space O(n)");
    }
}

File: CPP/data_structures/graphs/graph/disjoint_set_union.cpp
/**
 * @file disjoint_set_union.cpp
 * @brief Simple and optimized implementation of Disjoint Set Union (Union-Find)
 */

#include <iostream>
#include <vector>
#include <numeric> // for std::iota

// ---------------------- DSU Class ----------------------
class DSU {
private:
    std::vector<int> parent;          // parent[i] = parent of node i
    std::vector<int> component_size;  // size of each connected component
    int num_components;               // total number of disjoint sets

public:
    // Constructor: initializes DSU with 'n' elements (0 to n-1)
    DSU(int n) {
        num_components = n;
        parent.resize(n);
        component_size.resize(n);

        std::iota(parent.begin(), parent.end(), 0); // parent[i] = i
        std::fill(component_size.begin(), component_size.end(), 1); // all sizes = 1
    }

    // Find operation with path compression
    int find(int i) {
        if (parent[i] == i) return i;       // if i is its own parent → root
        return parent[i] = find(parent[i]); // path compression step
    }

    // Union operation with union by size
    bool unite(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA == rootB) return false; // already in same set

        // attach smaller tree to larger tree
        if (component_size[rootA] < component_size[rootB])
            std::swap(rootA, rootB);

        parent[rootB] = rootA;
        component_size[rootA] += component_size[rootB];
        num_components--;
        return true;
    }

    // Check if two elements belong to the same set
    bool inSameSet(int a, int b) {
        return find(a) == find(b);
    }

    // Get size of the set containing element 'i'
    int getComponentSize(int i) {
        return component_size[find(i)];
    }

    // Get total number of disjoint sets
    int getComponentCount() {
        return num_components;
    }
};

// ---------------------- Main Function ----------------------
int main() {
    DSU dsu(5); // 5 elements → {0}, {1}, {2}, {3}, {4}

    // Merge some sets
    dsu.unite(0, 1);
    dsu.unite(1, 2);
    dsu.unite(3, 4);

    // Check relationships
    std::cout << "Are 0 and 2 in same set? " 
              << (dsu.inSameSet(0, 2) ? "Yes" : "No") << "\n";

    std::cout << "Are 1 and 3 in same set? " 
              << (dsu.inSameSet(1, 3) ? "Yes" : "No") << "\n";

    // Show component sizes
    std::cout << "Size of component containing 0: " 
              << dsu.getComponentSize(0) << "\n";

    // Total number of disjoint sets
    std::cout << "Total components: " 
              << dsu.getComponentCount() << "\n";

    return 0;
}
